src/cached_object_store.rs

```diff
@@ -762,6 +817,286 @@ impl LocalCacheEntry for FsCacheEntry {
+        let scan_task_handle = tokio::spawn(inner.clone().scan_entries(true));
+        self.scan_task_handle.set(scan_task_handle).ok();
+
+        // start the background evictor task, it'll be triggered whenever a new cache entry is added
+        let evict_task_handle = tokio::spawn(Self::background_evict(rx, inner));
+        self.evict_task_handle.set(evict_task_handle).ok();
+    }
+
+    async fn started(&self) -> bool {
+        self.rx.lock().await.is_none()
+    }
+
+    async fn background_evict(
+        mut rx: tokio::sync::mpsc::Receiver<(std::path::PathBuf, usize)>,
+        inner: Arc<FsCacheEvictorInner>,
+    ) {
+        while let Some((path, bytes)) = rx.recv().await {
+            inner
+                .track_new_entry(path, bytes, SystemTime::now(), true)
+                .await;
+        }
+    }
+
+    pub async fn maybe_evict(&self, path: std::path::PathBuf, bytes: usize) {
+        if !self.started().await {
+            return;
+        }
+
+        self.tx.send((path, bytes)).await.ok();
+    }
+}
+
+/// FsCacheEvictorInner manages the cache entries in an in-memory trie, and evict the cache entries
+/// when the cache size exceeds the limit. it uses a pick-of-2 strategy to approximate LRU, and evict
+/// the older file when the cache size exceeds the limit.
+///
+/// On start up, FsCacheEvictorInner will scan the cache folder to load the cache files into the in-memory
+/// trie cache_entries. This loading process is interleaved with the maybe_evict is being called, so the
+/// cache entries should be wrapped with Mutex<_>.
+#[derive(Debug)]
+struct FsCacheEvictorInner {
+    root_folder: std::path::PathBuf,
+    batch_factor: usize,
+    max_cache_size_bytes: usize,
+    cache_entries: Mutex<Trie<std::path::PathBuf, (SystemTime, usize)>>,
+    cache_size_bytes: AtomicU64,
+    db_stats: Arc<DbStats>,
+}
+
+impl FsCacheEvictorInner {
+    pub fn new(
+        root_folder: std::path::PathBuf,
+        max_cache_size_bytes: usize,
+        db_stats: Arc<DbStats>,
+    ) -> Self {
+        Self {
+            root_folder,
+            batch_factor: 10,
+            max_cache_size_bytes,
+            cache_entries: Mutex::new(Trie::new()),
+            cache_size_bytes: AtomicU64::new(0_u64),
+            db_stats,
+        }
+    }
+
+    // scan the cache folder, and load the cache entries into the in memory trie cache_entries.
+    // this function is only called on start up, and it's expected to runned interleavely with
+    // maybe_evict is being called.
+    pub async fn scan_entries(self: Arc<Self>, evict: bool) {
+        // walk the cache folder, record the files and their last access time into the cache_entries
+        let iter = WalkDir::new(&self.root_folder).into_iter();
+        for entry in iter {
+            let entry = match entry {
+                Ok(entry) => entry,
+                Err(err) => {
+                    warn!("evictor: failed to walk the cache folder: {}", err);
```

Comment:
should we track a follow-up task to fail the db if we hit errors when scanning?

Reply:
i think a `panic` makes sense here, but i am not very sure about what's the possible reasons that really triggers this error.

if this error actually occurs occasionally, i guess the error is also tolerable? some entries might be missed during the scan, might makes eviction not accurate on checking the max cache size. however, this error is able to be auto recovered with the intervally rescan approach like what juicefs did. we did not put the rescan an intervally task as discussed in https://github.com/slatedb/slatedb/pull/227#discussion_r1785386992 .

i found that beside checking the max cache size, juicefs also detects the disks free ratio, which also helps mitigate this issue. i left a TODO about this in the code, how about raising an issue about checking the free space ratio of the disk?

Reply:
I think it's hard to predict when we might see errors. IMO it's easiest to deal with if we start by assuming that we won't see failures, and put the db in an error state when we do. Then, we can try to be smarter about it on a case-by-case basis (e.g. dealing with out-of-disk by evicting data). Otherwise it's hard to reason about what state the system is in (if it's just swallowing errors). I'll file a follow-up ticket.

Reply:
>  it's easiest to deal with if we start by assuming that we won't see failures, and put the db in an error state when we do.

+1

Reply:
I guess there might be these possible errors here:

1. deleted in a race condition from other tasks
2. permission error, accidentally misconfigured the permission wrong on these folders
3. fs corruption

On case (1), this could be considered as a tolerable error, as we did not lose any tracked file.

On cases (2) and (3), these are considered as unrecoverable; the program could not resolve the environment error by itself. It could choose:

1. fail early & fail loudly
2. tell the human by warnings or metric alerts, and tell humans to resolve these errors

Here, I kinda prefer to tell warnings / metrics to inform human & make auto-recovery in an eventual consistency manner.

The object store cache is designed not to strongly depend on a healthy fs. The goal of it is, even if your fs got corrupted, you can still fall back the requests to object storage as they are. So it feels not very comfortable for me to trigger unrecoverable error for user when it got any unexpected error in the evictor.

Failing fast & failing loudly is preferable when you encounter an inconsistent state of the program; we can stop the program to avoid things go even worse. But in this case, it's not considered as an inconsistent state internally but more like an incorrect environment config, and is considered better to not affect the data path of the program when your local fs is corrupted or misconfigured.

As above, all what it could do is tell human to fix the environment. If we tag an error state in the db, at last we still needs to choose how to handle this error between: 1. fail fast if you think this error is intolerable; 2. print warning & add metrics if we think the program can continue run but better to tell the human there're things need to be fixed.

In the sense of auto-recovery, the JuiceFS's intervally rescan approach may help here. If a human accidentally misconfigured the permission of a folder, the program could finally rescan these files after the human fixed the folder permission correctly in a eventual consistency manner.

I think "eventual consistency" here might be not a bad idea, as we do not nesscarily require a 1:1 strong consistent mapping of the files in the local fs & the in-memory map to make eviction works. If the cache files (or even some garbage files written by others) in the fs could be finally tracked by the evictor, we could finally triggers evictions to reclaim storage spaces.

@rodesai @criccomini WDYT?

Reply:
Makes sense especially given that, as you say, this is a cache and so we can handle some of these errors without failing the whole db. The drawback is that it makes the module trickier to reason about, because the number of possible states is increased (e.g. you now have some files on disk that are maybe not tracked by the evictor, there's retry/recovery logic). This also increases the likelihood of bugs. So usually my preference is to just fail hard unless there's a strong benefit from adding recovery logic (the underlying issue is common and likely to be rectified automatically without human intervention, or as is the case here the program can continue correctly but sub-optimally). In this case I'm ok with it as long as we're sure we won't ever return bad data from the cache, which we can ensure by checksumming the on-disk data - is there an issue open for this?

> I guess there might be these possible errors here:
deleted in a race condition from other tasks
permission error, accidentally misconfigured the permission wrong on these folders
fs corruption
On case (1), this could be considered as a tolerable error, as we did not lose any tracked file.
On cases (2) and (3), these are considered as unrecoverable; the program could not resolve the environment error by itself. It could choose:
fail early & fail loudly
tell the human by warnings or metric alerts, and tell humans to resolve these errors

There's also the possibility of network errors for attached storage in the cloud. I think this presents as a failed disk drive though, so it's not recoverable. It should be fine to handle it transparently and notify the user via logs/metrics.

The only case I disagree with is fs corruption. I think in cases where the data itself looks wrong we should treat it as a bad program state and fail hard and fast so that it's noticed quickly and any potential follow-on effects are minimized.

Reply:
sorry my word was inaccurate in my comment, my intention is to describe tolerating "fs error" like the mounted storage failure not data corruption, it's expected to be handled transparently in the current design.

when we talk about fs corruption, it's more likely about data being corrupted, and it's 100% a must to fail loudly with a checksum error.

this is the issue related to checksums: https://github.com/slatedb/slatedb/issues/173

Reply:
> The drawback is that it makes the module trickier to reason about, because the number of possible states is increased (e.g. you now have some files on disk that are maybe not tracked by the evictor, there's retry/recovery logic). This also increases the likelihood of bugs.

imho, things might get easier when we can regard the in-memory map as eventually consistent with the on-disk files. Things can finally become idempotent in the normal workflow, and we don't need to consider it as complex retry & recover logic.

i'll add an intervally rescan logic in this pr like juicefs did in https://github.com/juicedata/juicefs/blob/main/pkg/chunk/disk_cache.go#L367 , the change will be very trivial.

Reply:
added a intervally rescan mechanism in 1e60995